client/src/main/java/org/evosuite/ga/archive/MIOArchive.java

```diff
@@ -0,0 +1,575 @@
+ */
+package org.evosuite.ga.archive;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.evosuite.Properties;
+import org.evosuite.ga.Chromosome;
+import org.evosuite.ga.FitnessFunction;
+import org.evosuite.testcase.TestCase;
+import org.evosuite.testcase.TestFitnessFunction;
+import org.evosuite.testcase.execution.ExecutionResult;
+import org.evosuite.testsuite.TestSuiteChromosome;
+import org.evosuite.utils.Randomness;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of the archive described in the 'Many Independent Objective (MIO) Algorithm for
+ * Test Suite Generation' paper.
+ * 
+ * @author José Campos
+ */
+public class MIOArchive<F extends TestFitnessFunction, T extends TestCase> extends Archive<F, T> {
+
+  private static final long serialVersionUID = -6100903230303784634L;
+
+  private static final Logger logger = LoggerFactory.getLogger(MIOArchive.class);
+
+  /**
+   * Map used to store all targets (keys of the map) and the corresponding covering solutions
+   * (values of the map)
+   **/
+  protected final Map<F, Population> archive = new LinkedHashMap<F, Population>();
+
+  public static final MIOArchive<TestFitnessFunction, TestCase> instance =
+      new MIOArchive<TestFitnessFunction, TestCase>();
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void addTarget(F target) {
+    assert target != null;
+
+    if (!this.archive.containsKey(target)) {
+      logger.debug("Registering new target '" + target + "'");
+      this.archive.put(target, new Population(Properties.NUMBER_OF_TESTS_PER_TARGET));
+    }
+
+    this.registerNonCoveredTargetOfAMethod(target);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @SuppressWarnings("unchecked")
+  @Override
+  public void updateArchive(F target, ExecutionResult executionResult, double fitnessValue) {
+    assert target != null;
+    assert this.archive.containsKey(target);
+    assert fitnessValue >= 0.0;
+
+    ExecutionResult executionResultClone = executionResult.clone();
+    T solutionClone = (T) executionResultClone.test.clone(); // in case executionResult.clone() has
+                                                             // not cloned the test
+    executionResultClone.setTest(solutionClone);
+    // remove all statements after an exception
+    if (!executionResultClone.noThrownExceptions()) {
+      solutionClone.chop(executionResultClone.getFirstPositionOfThrownException() + 1);
+    }
+
+    boolean isNewCoveredTarget = this.archive.get(target).addSolution(
+        1.0 - (fitnessValue <= 1.0 ? fitnessValue : FitnessFunction.normalize(fitnessValue)),
```

Comment:
unsure if this is correct... if you are normalizing, shouldn't you always normalize?
I mean, what happen if a normalized 1.x gets a lower value than a non-normalized 0.9999

Reply:
Yes, that is definitively wrong. It should always normalize.